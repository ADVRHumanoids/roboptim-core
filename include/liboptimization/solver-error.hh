// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the SolverError class.
 */

#ifndef OPTIMIZATION_SOLVER_ERROR_HH
# define OPTIMIZATION_SOLVER_ERROR_HH
# include <stdexcept>

namespace optimization
{
  /// \brief Base error.
  /// All other errors inherits this class.
  struct SolverError : public std::runtime_error
  {
    explicit SolverError (const std::string& arg) throw ();
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_SOLVER_ERROR_HH
