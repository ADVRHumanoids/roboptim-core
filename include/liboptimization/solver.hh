// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Solver class.
 */

#ifndef OPTIMIZATION_SOLVER_HH
# define OPTIMIZATION_SOLVER_HH
# include <boost/static_assert.hpp>
# include <boost/type_traits/is_base_of.hpp>

# include <liboptimization/fwd.hh>
# include <liboptimization/function.hh>
# include <liboptimization/problem.hh>
# include <liboptimization/generic-solver.hh>

namespace optimization
{
  /// Define a solver for a class of problem.
  template <typename F, typename C>
  class Solver : public GenericSolver
  {
    BOOST_STATIC_ASSERT((boost::is_base_of<Function, F>::value));
  public:

    typedef Problem<F, C> problem_t;

    explicit Solver (const problem_t&) throw ();

    template <typename F_, typename C_>
    explicit Solver (const Problem<F_, C_>&) throw ();

    ~Solver () throw ();

    const problem_t& problem () const throw ();
  protected:
    const problem_t problem_;
  };



}; // end of namespace optimization

# include <liboptimization/solver.hxx>
#endif //! OPTIMIZATION_SOLVER_HH
