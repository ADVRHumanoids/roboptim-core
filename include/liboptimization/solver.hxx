// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Implementation of the Solver class.
 */

#ifndef OPTIMIZATION_SOLVER_HXX
# define OPTIMIZATION_SOLVER_HXX

namespace optimization
{
  template <typename F, typename C>
  Solver<F, C>::Solver (const problem_t& pb) throw ()
    : GenericSolver (),
      problem_ (pb)
  {
  }

  template <typename F, typename C>
  template <typename F_, typename C_>
  Solver<F, C>::Solver (const Problem<F_, C_>& pb) throw ()
    : GenericSolver (),
      problem_ (pb)
  {
  }

  template <typename F, typename C>
  Solver<F, C>::~Solver () throw ()
  {
  }

  template <typename F, typename C>
  const typename Solver<F, C>::problem_t&
  Solver<F, C>::problem () const throw ()
  {
    return problem_;
  }

}; // end of namespace optimization

#endif //! OPTIMIZATION_SOLVER_HH
