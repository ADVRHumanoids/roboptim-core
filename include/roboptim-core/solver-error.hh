// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the SolverError class.
 */

#ifndef OPTIMIZATION_SOLVER_ERROR_HH
# define OPTIMIZATION_SOLVER_ERROR_HH
# include <iostream>
# include <stdexcept>

namespace roboptim
{
  /// \brief Base error.
  /// All other errors inherits this class.
  class SolverError : public std::runtime_error
  {
  public:
    explicit SolverError (const std::string& arg) throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  };

  std::ostream& operator<< (std::ostream&, const SolverError&);
} // end of namespace roboptim

#endif //! OPTIMIZATION_SOLVER_ERROR_HH
