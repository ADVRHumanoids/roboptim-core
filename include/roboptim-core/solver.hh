// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

/**
 * \brief Declaration of the Solver class.
 */

#ifndef OPTIMIZATION_SOLVER_HH
# define OPTIMIZATION_SOLVER_HH
# include <boost/static_assert.hpp>
# include <boost/type_traits/is_base_of.hpp>

# include <roboptim-core/fwd.hh>
# include <roboptim-core/function.hh>
# include <roboptim-core/problem.hh>
# include <roboptim-core/generic-solver.hh>

namespace roboptim
{
  /// Define a solver for a class of problem.
  template <typename F, typename C>
  class Solver : public GenericSolver
  {
    BOOST_STATIC_ASSERT((boost::is_base_of<Function, F>::value));
  public:

    typedef Problem<F, C> problem_t;

    explicit Solver (const problem_t&) throw ();

    template <typename F_, typename C_>
    explicit Solver (const Problem<F_, C_>&) throw ();

    ~Solver () throw ();

    const problem_t& problem () const throw ();

    virtual std::ostream& print (std::ostream&) const throw ();
  protected:
    const problem_t problem_;
  };

} // end of namespace roboptim

# include <roboptim-core/solver.hxx>
#endif //! OPTIMIZATION_SOLVER_HH
