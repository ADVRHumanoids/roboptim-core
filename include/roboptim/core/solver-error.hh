// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_SOLVER_ERROR_HH
# define ROBOPTIM_CORE_SOLVER_ERROR_HH
# include <iostream>
# include <stdexcept>

namespace roboptim
{
  /// \addtogroup roboptim_problem
  /// @{

  /// \brief Base exception class for solving errors.
  /// All other exceptions classes concerning the optimization
  /// process should inherits this class.
  class SolverError : public std::runtime_error
  {
  public:
    /// \brief Instantiate an error from an error message.
    /// \param arg error message.
    explicit SolverError (const std::string& arg) throw ();

    /// \brief Display the error on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();
  };

  /// @}

  /// \brief Override operator<< to handle error display.
  ///
  /// \param o output stream used for display
  /// \param e error to be displayed
  /// \return output stream
  std::ostream& operator<< (std::ostream& o, const SolverError& e);
} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_SOLVER_ERROR_HH
