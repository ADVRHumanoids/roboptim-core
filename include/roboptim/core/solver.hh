// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_SOLVER_HH
# define ROBOPTIM_CORE_SOLVER_HH
# include <roboptim/core/sys.hh>
# include <roboptim/core/debug.hh>

# include <map>
# include <string>

# include <boost/function.hpp>
# include <boost/mpl/assert.hpp>
# include <boost/mpl/logical.hpp>
# include <boost/type_traits/is_base_of.hpp>
# include <boost/variant/variant.hpp>
# include <boost/variant/get.hpp>

# include <roboptim/core/fwd.hh>
# include <roboptim/core/function.hh>
# include <roboptim/core/problem.hh>
# include <roboptim/core/generic-solver.hh>

namespace roboptim
{
  /// \addtogroup roboptim_problem
  /// @{

  /// \brief Parameters type.
  struct Parameter
  {
    /// \brief Allowed types for parameters.
    typedef boost::variant<Function::value_type,
			   int, std::string> parameterValues_t;

    /// \brief Parameter description (for humans).
    std::string description;

    /// \brief Value.
    parameterValues_t value;
  };

  /// \brief Solver for a specific problem class.
  ///
  /// This class is parametrized by two types:
  /// the cost function type and the constraints type.
  ///
  /// Solver classes are immutable, the problem can
  /// not be changed after the class instantiation.
  ///
  /// \tparam F cost function type
  /// \tparam C constraints functions type
  /// \pre F is a subtype of Function
  template <typename F, typename C>
  class Solver : public GenericSolver
  {
    BOOST_MPL_ASSERT((boost::mpl::or_<boost::is_base_of<Function, F>,
		      boost::is_base_of<SparseFunction, F> >));
  public:
    /// \brief Solver problem type.
    ///
    /// The solver can solve problems of this type.
    /// If another kind of problem is given, a conversion will be
    /// required.
    typedef Problem<F, C> problem_t;

    /// \brief Import vector type from cost function
    typedef typename F::vector_t vector_t;

    /// \brief Map of parameters.
    typedef std::map<std::string, Parameter> parameters_t;

    /// Per-iteration callback type
    ///
    /// Callback parameters:
    /// \li x is the current considered point,
    /// \li problem is a (constant) reference to the problem
    ///
    typedef boost::function<void (const vector_t& x,
				  const problem_t& problem)> callback_t;


    /// \brief Instantiate a solver from a problem.
    ///
    /// \param problem problem that should be solved
    explicit Solver (const problem_t& problem) throw ();


    /// \brief Instantiate a solver from a problem in a different problem class.
    ///
    /// This constructor is called when the problem cost function or/and
    /// constraints type does not match solver's types.
    ///
    /// This is only possible if the problem provides too much information
    /// compared to the solver requirements:
    /// if the problem contains twice derivable function and the solver requires
    /// only derivable function, it will work however the opposite will fail.
    /// Problem compatibility is known at compile-time, so the failure will be
    /// at compile-time.
    ///
    /// \tparam F_ original cost function type
    /// \tparam C_ original constraints functions type
    /// \param problem problem that should be solved
    template <typename F_, typename C_>
    explicit Solver (const Problem<F_, C_>& problem) throw ();

    virtual ~Solver () throw ();

    /// \brief Retrieve the problem.
    /// \return problem this solver is solving
    const problem_t& problem () const throw ();

    /// \name Parameters
    /// \{
    const parameters_t& parameters () const throw ();
    parameters_t& parameters () throw ();

    template <typename T>
    const T& getParameter (const std::string& key) const;
    /// \}

    /// \brief Set the per-iteration callback.
    ///
    /// The per-iteration callback is a callback called each time one
    /// iteration of the optimization process is finished.
    ///
    /// Not all the solvers support such a callback so this method may
    /// throw a std::runtime_error to let you know this feature is
    /// unsupported.
    virtual void
    setIterationCallback (callback_t /*callback*/) throw (std::runtime_error)
    {
      throw std::runtime_error
	("iteration callback is not supported by this solver");
    }

    /// \brief Display the solver on the specified output stream.
    ///
    /// \param o output stream used for display
    /// \return output stream
    virtual std::ostream& print (std::ostream&) const throw ();
  protected:
    /// \brief Problem that will be solved.
    const problem_t problem_;

    /// \brief Solver parameters (run-time configuration).
    parameters_t parameters_;

    /// \brief Pointer to function logger (see log4cxx documentation).
    static log4cxx::LoggerPtr logger;
  };

  /// \brief Override operator<< to display ``parameters'' objects.
  ///
  /// \param o output stream used for display
  /// \param ns NoSolution object, ignored
  /// \return output stream
  ROBOPTIM_DLLAPI std::ostream&
  operator<< (std::ostream& o, const Parameter& parameter);
  /// @}

} // end of namespace roboptim
# include <roboptim/core/solver.hxx>
#endif //! ROBOPTIM_CORE_SOLVER_HH
