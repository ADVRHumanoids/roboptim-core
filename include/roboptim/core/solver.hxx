// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_SOLVER_HXX
# define ROBOPTIM_CORE_SOLVER_HXX
# include <boost/foreach.hpp>
# include <roboptim/core/io.hh>

namespace roboptim
{
  template <typename F, typename C>
  Solver<F, C>::Solver (const problem_t& pb) throw ()
    : GenericSolver (),
      problem_ (pb)
  {
  }

  template <typename F, typename C>
  template <typename F_, typename C_>
  Solver<F, C>::Solver (const Problem<F_, C_>& pb) throw ()
    : GenericSolver (),
      problem_ (pb)
  {
  }

  template <typename F, typename C>
  Solver<F, C>::~Solver () throw ()
  {
  }

  template <typename F, typename C>
  const typename Solver<F, C>::problem_t&
  Solver<F, C>::problem () const throw ()
  {
    return problem_;
  }

  template <typename F, typename C>
  const typename Solver<F, C>::parameters_t&
  Solver<F, C>::parameters () const throw ()
  {
    return parameters_;
  }

  template <typename F, typename C>
  typename Solver<F, C>::parameters_t&
  Solver<F, C>::parameters () throw ()
  {
    return parameters_;
  }

  template <typename F, typename C>
  template <typename T>
  const T&
  Solver<F, C>::getParameter (const std::string& key) const
  {
    parameters_t::const_iterator it = parameters_.find (key);
    return boost::get<T> (it->second.value);
  }


  template <typename F, typename C>
  std::ostream&
  Solver<F, C>::print (std::ostream& o) const throw ()
  {
    o << incindent << "Solver:" << iendl
      << this->problem_ << iendl
      << "Parameters:" << incindent << iendl;

    typedef const std::pair<const std::string, Parameter> const_iterator_t;
    BOOST_FOREACH (const_iterator_t& it, parameters_)
      o << it.first << " (" << it.second.description << ")" << ": "
	<< it.second.value << iendl;
    o << decindent << decindent;
    return o;
  }
} // end of namespace roboptim
#endif //! ROBOPTIM_CORE_SOLVER_HH
