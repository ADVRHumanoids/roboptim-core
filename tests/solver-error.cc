// Copyright (C) 2016 by Benjamin Chr√©tien, CNRS-AIST JRL.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include "shared-tests/fixture.hh"

#include <iostream>

#include <boost/shared_ptr.hpp>
#include <boost/make_shared.hpp>

#include <roboptim/core/io.hh>
#include <roboptim/core/solver.hh>
#include <roboptim/core/solver-error.hh>

using namespace roboptim;

boost::shared_ptr<boost::test_tools::output_test_stream> output;

// Define a simple function.
struct F1 : public Function
{
  F1 () : Function (4, 1, "a + b + c + d")
  {}

  void impl_compute (result_ref res,
                     const_argument_ref x) const
  {
    res (0) = x[0] + x[1] + x[2] + x[3];
  }

  // No gradient, hessian.
};

class DummyErrorSolver : public Solver<EigenMatrixDense>
{
public:
  /// Define parent's type.
  typedef Solver<EigenMatrixDense> parent_t;
  typedef parent_t::solverState_t solverState_t;
  typedef parent_t::callback_t callback_t;

  DummyErrorSolver (const problem_t& pb)
    : parent_t (pb)
  {
  }

  ~DummyErrorSolver ()
  {
  }

  void solve ()
  {
    (*output) << "solve ()" << std::endl;

    result_ = SolverError ("The dummy solver always fail.\n"
                           "It also prints a dummy multiline error message.");
  }
};

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

BOOST_AUTO_TEST_CASE (solver_error)
{
  output = retrievePattern ("solver-error");

  typedef DummyErrorSolver solver_t;
  boost::shared_ptr<F1> f = boost::make_shared<F1> ();
  solver_t::problem_t pb (f);

  boost::shared_ptr<solver_t> solver (new solver_t (pb));
  solver->solve ();

  (*output) << *solver << std::endl;

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
