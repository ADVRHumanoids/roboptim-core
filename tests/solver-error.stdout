solve ()
Solver:
  Problem:
    a + b + c + d (not differentiable)
    Arguments bounds: (-1.#INF, 1.#INF), (-1.#INF, 1.#INF), (-1.#INF, 1.#INF), (-1.#INF, 1.#INF)
    Arguments scaling: 1, 1, 1, 1
    No constraints.
    No starting point.
    Infinity value (for all functions): 1.#INF
  Result:
    Solver error:
      The dummy solver always fail.
      It also prints a dummy multiline error message.

